aws-resources-cfn-template.yaml

```diff
@@ -93,3 +118,215 @@ Resources:
+            Resource:
+              - !GetAtt AssetBucket.Arn
+              - !Sub ${AssetBucket.Arn}/*
+            Condition:
+              Bool:
+                'aws:SecureTransport': false
+          - Action: "s3:GetObject"
+            Effect: Allow
+            Principal:
+              CanonicalUser: !GetAtt AssetBucketOAI.S3CanonicalUserId
+            Resource: !Sub ${AssetBucket.Arn}/*
+        Version: 2012-10-17
+
+  PublicAssetDistribution:
+    Type: "AWS::CloudFront::Distribution"
+    Properties:
+      DistributionConfig:
+        DefaultCacheBehavior:
+          CachePolicyId: 658327ea-f89d-4fab-a63d-7e88639e58f6
+          Compress: true
+          TargetOriginId: IotAppFrontEndAssetDistributionOrigin
+          ViewerProtocolPolicy: redirect-to-https
+        CacheBehaviors:
+          - PathPattern: aws-resources.js
+            TargetOriginId: IotAppS3BucketOrigin
+            ViewerProtocolPolicy: redirect-to-https
+            AllowedMethods:
+              - GET
+              - HEAD
+            CachePolicyId: 658327ea-f89d-4fab-a63d-7e88639e58f6
+        DefaultRootObject: index.html
+        Enabled: true
+        HttpVersion: http2
+        IPV6Enabled: true
+        Origins:
+          - DomainName: !Ref FrontEndAssetDistribution
+            Id: IotAppFrontEndAssetDistributionOrigin
+            CustomOriginConfig:
+              OriginProtocolPolicy: https-only
+              OriginSSLProtocols:
+                - TLSv1.2
+          - DomainName: !GetAtt AssetBucket.RegionalDomainName
+            Id: IotAppS3BucketOrigin
+            S3OriginConfig:
+              OriginAccessIdentity: !Sub origin-access-identity/cloudfront/${AssetBucketOAI}
+
+  AwsResourcesFile:
+    Type: Custom::AwsResourcesFile
+    Properties:
+      ServiceToken: !GetAtt S3PutObjectFunction.Arn
+      Bucket: !Ref AssetBucket
+      Body: !Sub window.awsResources={"amplifyConfiguration":{"Auth":{"identityPoolId":"${IdentityPool}","region":"${AWS::Region}","userPoolId":"${UserPool}","userPoolWebClientId":"${UserPoolClientWeb}"}},"coreServer":{"endpoint":"https://${CoreService.ServiceUrl}"}};
+      Key: aws-resources.js
+
+  S3PutObjectFunction:
+    Type: AWS::Lambda::Function
+    Properties:
+      Handler: index.handler
+      Role: !GetAtt S3PutObjectFunctionRole.Arn
+      Timeout: 360
+      Runtime: nodejs16.x
+      Code:
+        ZipFile: |
+          const aws = require('aws-sdk');
+          const cfnresponse = require('./cfn-response');
+
+          const s3 = new aws.S3();
+
+          exports.handler = (event, context) => {
+            // Get the request event info
+            const requestType = event.RequestType;
+            console.log('The event type is: ', requestType);
+            console.log('The resource properties is: ', event.ResourceProperties);
+            const { Body, Bucket, Key } = event.ResourceProperties;
+
+            if (requestType === 'Create' || requestType === 'Update') {
```

Comment:
Is request type of union of string literals? If so, with TypeScript 5, there's a compiler flag for exhaustive switches, which will ensure all cases are handled. Pretty cool!